src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java

```diff
@@ -0,0 +1,102 @@
+import com.commercetools.sync.commons.models.GraphQlQueryResources;
+import com.commercetools.sync.commons.models.ResourceIdsGraphQlRequest;
+import com.commercetools.sync.commons.models.ResourceKeyId;
+import com.commercetools.sync.commons.utils.ChunkUtils;
+import io.sphere.sdk.client.SphereClient;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nonnull;
+
+public class BaseTransformServiceImpl {
+  /*
+   * An id is a 36 characters long string. (i.e: 53c4a8b4-754f-4b95-b6f2-3e1e70e3d0c3) We
+   * chunk them in 300 ids, we will have a query around 11.000 characters. Above this size it
+   * could return - Error 413 (Request Entity Too Large)
+   */
+  public static final int CHUNK_SIZE = 300;
+  public static final String BLANK_KEY_VALUE = "KEY_VALUE_EITHER_NULL_OR_BLANK";
+  protected final Map<String, String> referenceIdToKeyCache;
+
+  private final SphereClient ctpClient;
+
+  protected BaseTransformServiceImpl(
+      @Nonnull final SphereClient ctpClient,
+      @Nonnull final Map<String, String> referenceIdToKeyCache) {
+    this.ctpClient = ctpClient;
+    this.referenceIdToKeyCache = referenceIdToKeyCache;
+  }
+
+  protected SphereClient getCtpClient() {
+    return ctpClient;
+  }
+
+  protected CompletableFuture<Void> fetchAndFillReferenceIdToKeyCache(
+      @Nonnull final Set<String> ids, @Nonnull final GraphQlQueryResources requestType) {
+
+    final Set<String> nonCachedReferenceIds = getNonCachedReferenceIds(ids);
+    if (nonCachedReferenceIds.isEmpty()) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    final List<List<String>> chunkedIds = ChunkUtils.chunk(nonCachedReferenceIds, CHUNK_SIZE);
+
+    List<ResourceIdsGraphQlRequest> collectedRequests =
+        createResourceIdsGraphQlRequests(chunkedIds, requestType);
+
+    return ChunkUtils.executeChunks(getCtpClient(), collectedRequests)
+        .thenApply(ChunkUtils::flattenGraphQLBaseResults)
+        .thenCompose(
+            results -> {
+              try {
+                cacheResourceReferenceKeys(results);
+              } catch (ReferenceTransformException referenceTransformException) {
+                return exceptionallyCompletedFuture(referenceTransformException);
+              }
+              return CompletableFuture.completedFuture(null);
+            });
+  }
+
+  @Nonnull
+  protected List<ResourceIdsGraphQlRequest> createResourceIdsGraphQlRequests(
+      @Nonnull final List<List<String>> chunkedIds,
+      @Nonnull final GraphQlQueryResources resourceType) {
+    return chunkedIds.stream()
+        .map(chunk -> new ResourceIdsGraphQlRequest(new HashSet<>(chunk), resourceType))
+        .collect(toList());
+  }
+
+  @Nonnull
+  protected Set<String> getNonCachedReferenceIds(@Nonnull final Set<String> referenceIds) {
+    return referenceIds.stream()
```

Comment:
Here we need to filter out the keys not exist but have a placeholder, KEY_IS_NOT_SET_PLACE_HOLDER. 
Otherwise, when the key is added to the project it would not be synced and the exceptions will not be resolved.

Could you check for other code if `getNonCachedReferenceIds` is overridden in subclasses? Also please try to check document reference for custom objects, as it's not stored as a simple key but the combination of container and key (("id", "co-container|co-key"))

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/697/commits/ffe6f19db27aa206410807c5e56943283370b0c3
I am checking for custom Objects.